### PR TITLE
[Improvement] Set heartBeatExecutorService as daemon thread

### PR DESCRIPTION
**Pull request by smallzhongfeng**
### What changes were proposed in this pull request?

When the main thread of the shuffleServer exits, the heartbeat thread should exit as well.

### Why are the changes needed?

More formal.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

No need.

### CODE REVIEW

**Comment by smallzhongfeng** [comment]
cc @jerqi

**Comment by jerqi** [comment]
Why do we need this patch?

**Comment by smallzhongfeng** [comment]
I'm not sure whether the main thread of the `shuffleServer` will exit unexpectedly, but is it meaningful to set the heartbeat scheduling thread as the daemon thread.

**Comment by jerqi** [comment]
> I'm not sure whether the main thread of the `shuffleServer` will exit unexpectedly, but is it meaningful to set the heartbeat scheduling thread as the daemon thread.

Ok, have you checked similar problems in other code in our project? Could you fix them together?

**Comment by smallzhongfeng** [comment]
The same as `ExecutorService executor = Executors.newSingleThreadExecutor();` I think this one needs to be set to a daemon thread as well. What do you think?@jerqi

**Comment by jerqi** [comment]
> The same as `ExecutorService executor = Executors.newSingleThreadExecutor();` I think this one needs to be set to a daemon thread as well. What do you think?@jerqi

I think we should check all types ExecutorService. A similar pr https://github.com/apache/incubator-uniffle/commit/ba47aa017f67e681af7c311c4ef8578eef740d4b

**Comment by smallzhongfeng** [comment]
Reference pr you metioned, I set all unnecessary user threads as daemon.

File: client-mr/src/main/java/org/apache/hadoop/mapred/SortWriteBufferManager.java [hunk 1]
```diff
@@ -321,7 +321,7 @@ ShuffleBlockInfo createShuffleBlock(SortWriteBuffer wb) {
   }
 
   protected void sendCommit() {
-    ExecutorService executor = Executors.newSingleThreadExecutor();
+    ExecutorService executor = Executors.newSingleThreadExecutor(ThreadUtils.getThreadFactory("bufferManagerSend-%d"));
```
Comment (jerqi) [inline client-mr/src/main/java/org/apache/hadoop/mapred/SortWriteBufferManager.java:324]: This is a temporary thread pool. I think it's unnecessary to be a daemon thread pool.WDYT?

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 2]
```diff
@@ -244,7 +245,7 @@ protected void checkBlockSendResult(Set<Long> blockIds) throws RuntimeException
 
   @VisibleForTesting
   protected void sendCommit() {
-    ExecutorService executor = Executors.newSingleThreadExecutor();
+    ExecutorService executor = Executors.newSingleThreadExecutor(ThreadUtils.getThreadFactory("sendCommit-%d"));
```
Comment (jerqi) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:248]: This is a temporary thread pool. I think it's unnecessary to be a daemon thread pool.WDYT?

File: server/src/test/java/org/apache/uniffle/server/ShuffleServerMetricsTest.java [hunk 3]
```diff
@@ -150,7 +151,8 @@ public void testGrpcMetrics() throws Exception {
 
   @Test
   public void testServerMetricsConcurrently() throws Exception {
-    ExecutorService executorService = Executors.newFixedThreadPool(3);
+    ExecutorService executorService = Executors.newCachedThreadPool(
```
Comment (jerqi) [inline server/src/test/java/org/apache/uniffle/server/ShuffleServerMetricsTest.java:154]: This is a temporary thread pool. I think it's unnecessary to be a daemon thread pool.WDYT?

File: server/src/main/java/org/apache/uniffle/server/storage/LocalStorageManager.java [hunk 4]
```diff
@@ -81,7 +82,8 @@ public class LocalStorageManager extends SingleStorageManager {
     // We must make sure the order of `storageBasePaths` and `localStorages` is same, or some unit test may be fail
     CountDownLatch countDownLatch = new CountDownLatch(storageBasePaths.length);
     AtomicInteger successCount = new AtomicInteger();
-    ExecutorService executorService = Executors.newCachedThreadPool();
+    ExecutorService executorService = Executors.newCachedThreadPool(
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/storage/LocalStorageManager.java:85]: This is a temporary thread pool. I think it's unnecessary to be a daemon thread pool.WDYT?

File: client-mr/src/main/java/org/apache/hadoop/mapred/SortWriteBufferManager.java (hunk 1 above)
Comment (smallzhongfeng) [reply client-mr/src/main/java/org/apache/hadoop/mapred/SortWriteBufferManager.java:324]: Yes, I think I made a mistake here, but I think a better way would have been to generate this executor at this object initialization. @jerqi The same as below.

Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapred/SortWriteBufferManager.java:324]: MEMORY_LOCALFILE and MEMROY_HDFS don't call commit operation, so we use a temporary thread pool.

Comment (smallzhongfeng) [reply client-mr/src/main/java/org/apache/hadoop/mapred/SortWriteBufferManager.java:324]: Got it.

File: server/src/main/java/org/apache/uniffle/server/storage/LocalStorageManager.java (hunk 4 above)
Comment (smallzhongfeng) [reply server/src/main/java/org/apache/uniffle/server/storage/LocalStorageManager.java:85]: Would it be better to create the thread pool when the `LocalStorageManager` is initialized.

Comment (jerqi) [reply server/src/main/java/org/apache/uniffle/server/storage/LocalStorageManager.java]: We only use this thread pool to delete the files once when the server start. I think it's no need.

File: server/src/main/java/org/apache/uniffle/server/storage/LocalStorageManager.java [hunk 5]
```diff
@@ -61,6 +62,8 @@ public class LocalStorageManager extends SingleStorageManager {
   private final LocalStorageChecker checker;
   private List<LocalStorage> unCorruptedStorages = Lists.newArrayList();
   private final Set<String> corruptedStorages = Sets.newConcurrentHashSet();
+  private final ExecutorService executorService = Executors.newCachedThreadPool(
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/storage/LocalStorageManager.java:65]: We only use this thread pool to delete the files once when the server start. I think it's no need.

Comment (smallzhongfeng) [reply server/src/main/java/org/apache/uniffle/server/storage/LocalStorageManager.java]: Updated.

File: server/src/test/java/org/apache/uniffle/server/ShuffleServerMetricsTest.java [hunk 6]
```diff
@@ -150,7 +150,7 @@ public void testGrpcMetrics() throws Exception {
 
   @Test
   public void testServerMetricsConcurrently() throws Exception {
-    ExecutorService executorService = Executors.newFixedThreadPool(3);
+    ExecutorService executorService = Executors.newCachedThreadPool();
```
Comment (jerqi) [inline server/src/test/java/org/apache/uniffle/server/ShuffleServerMetricsTest.java:153]: Why do we modify this?

Comment (smallzhongfeng) [reply server/src/test/java/org/apache/uniffle/server/ShuffleServerMetricsTest.java:153]: IIRC, a cachedThreadPool should be faster if the number of threads started is known though this is just a test class.

Comment (smallzhongfeng) [reply server/src/test/java/org/apache/uniffle/server/ShuffleServerMetricsTest.java:153]: I'm not sure why these thread numbers are needed here.

Comment (smallzhongfeng) [reply server/src/test/java/org/apache/uniffle/server/ShuffleServerMetricsTest.java]: Got it.

**Review by jerqi — APPROVED** [review]
LGTM, thanks @smallzhongfeng

**Pull request merged**